File: watch/wearos/src/main/java/app/vescape/wear/WeatherReadout.kt
package app.vescape.wear

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.wear.compose.material.Icon
import androidx.wear.compose.material.MaterialTheme
import androidx.wear.compose.material.Text

/**
 * A temperature with its degree mark tucked against the last digit. The ° hangs on the digits as an
 * overlay, so it never shifts where the number itself lands: "21°" and "20" share one centre, and
 * the mark is a tail, not part of the column.
 *
 * @parity /watch/watchos/WeatherGlyph.swift `DegreeNumber`
 */
@Composable
internal fun DegreeTemp(value: Int, style: TextStyle, color: Color, modifier: Modifier = Modifier) {
    Box(modifier = modifier, contentAlignment = Alignment.Center) {
        Text(text = value.toString(), style = style, color = color)
        Text(
            text = "°",
            style = style,
            color = color,
            modifier = Modifier
                .align(Alignment.CenterEnd)
                .offset(x = style.fontSize.value.dp * DEGREE_TUCK),
        )
    }
}

/** Shift that leaves a hair of space between the last digit and the mark.
 *
 * @parity /watch/watchos/WeatherGlyph.swift `degreeTuck`
 */
private const val DEGREE_TUCK = 0.45f

/**
 * The forecast strip under the wall clock on the gauges page: condition glyph and temperature, with
 * the chance of rain stacked under them whenever there is one. Sized to disappear into the rim gap —
 * a rider glancing at speed should register it without it competing with a gauge, which is also why
 * rain goes below rather than beside: a wider strip would reach into the speed and duty gauges.
 *
 * Renders nothing until the phone has pushed a forecast, so the layout above the gauges is unchanged
 * on a phone too old to send one. Tapping opens [WeatherScreen].
 *
 * @parity /watch/watchos/WeatherReadout.swift
 */
@Composable
internal fun WeatherReadout(muted: Boolean, onClick: (() -> Unit)?, modifier: Modifier = Modifier) {
    val forecast = freshWeather() ?: return
    val iconColor = if (muted) DimText else weatherColor(forecast.icon)
    val textColor = if (muted) DimText else SecondaryText

    Column(
        // Null while another page has the screen: the readout is drawn over the pagers so the tap
        // can reach it at all, and a target up there would swallow drags meant for them. A disabled
        // `clickable` still installs a pointer-input node, so the modifier has to be absent, not off.
        modifier = modifier
            .then(if (onClick != null) Modifier.clickable(onClick = onClick) else Modifier)
            .padding(horizontal = 6.dp, vertical = 2.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Icon(
                painter = painterResource(weatherIconRes(forecast.icon)),
                contentDescription = forecast.label,
                tint = iconColor,
                modifier = Modifier.size(READOUT_ICON_SIZE),
            )
            DegreeTemp(
                value = forecast.temperatureC,
                style = WatchTypography.mono(MaterialTheme.typography.caption2.copy(fontSize = READOUT_FONT_SIZE)),
                color = textColor,
                modifier = Modifier.padding(start = 3.dp),
            )
        }
        if (forecast.precipitationProbability > 0) {
            // All blue, glyph and number alike: rain is one reading, not an icon with a label.
            val rainColor = if (muted) DimText else weatherColor("cloud-rain")
            Row(verticalAlignment = Alignment.CenterVertically) {
                Icon(
                    painter = painterResource(R.drawable.ic_ph_drop),
                    contentDescription = null,
                    tint = rainColor,
                    modifier = Modifier.size(READOUT_DROP_SIZE),
                )
                Text(
                    text = "${forecast.precipitationProbability}%",
                    style = WatchTypography.mono(MaterialTheme.typography.caption2.copy(fontSize = RAIN_FONT_SIZE)),
                    color = rainColor,
                    modifier = Modifier.padding(start = 1.dp),
                )
            }
        }
    }
}

private val READOUT_ICON_SIZE = 13.dp
private val READOUT_DROP_SIZE = 8.dp
private val READOUT_FONT_SIZE = 11.sp
private val RAIN_FONT_SIZE = 9.sp

File: watch/watchos/WeatherReadout.swift
import SwiftUI

/// Compact forecast above speed: condition and temperature, with rain below.
///
/// Renders nothing until the phone has pushed a forecast, so the gauges page is unchanged on a
/// phone that has not seen a GPS Fix yet. Tapping it opens the weather page.
///
/// Wear OS hangs this under its own wall clock, in the gap the rim arcs leave at the top of the
/// circle. watchOS puts the strip inside the top-left rim, above the speed number and to the left
/// of the system clock.
///
/// @parity /watch/wearos/src/main/java/app/vescape/wear/WeatherReadout.kt `WeatherReadout`
/// @platform-diff Wear OS places the strip under its own wall clock; on watchOS the system clock
///   sits at the upper right, so the strip sits above speed at the upper left.
struct WeatherReadout: View {
  let forecast: WatchWeather?
  var ambient: AmbientMode = .off
  /// Nil while another page owns the screen: a tap target here would otherwise swallow drags meant
  /// for the pagers.
  var onTap: (() -> Void)?

  var body: some View {
    if let forecast {
      strip(forecast)
    }
  }

  private func strip(_ forecast: WatchWeather) -> some View {
    let iconColor = ambient.active ? Palette.dimText : Palette.weather(forecast.icon)
    let textColor = ambient.skeleton(Palette.secondaryText)
    let rainColor = ambient.active ? Palette.dimText : Palette.weather("cloud-rain")

    return VStack(alignment: .trailing, spacing: 1) {
      HStack(spacing: 0) {
        WeatherGlyph(slug: forecast.icon, size: ICON_SIZE, color: iconColor)
        // Tucked °, so the temp's trailing edge is its last digit and the rain row below hangs
        // flush with it instead of clearing the mark's advance width.
        DegreeNumber(value: "\(forecast.temperatureC)", size: FONT_SIZE, color: textColor)
      }
      if forecast.precipitationProbability > 0 {
        // All blue, glyph and number alike: rain is one reading, not an icon with a label.
        HStack(spacing: 1) {
          Image(systemName: rainSymbol)
            .font(.system(size: DROP_SIZE))
          Text("\(forecast.precipitationProbability)%")
            .font(WatchTypography.mono(size: RAIN_FONT_SIZE))
            .monospacedDigit()
        }
        .foregroundStyle(rainColor)
      }
    }
    .fixedSize()
    .contentShape(Rectangle())
    // A disabled tap gesture still installs a gesture recognizer, so the modifier has to be absent
    // rather than switched off.
    .modifier(OptionalTap(action: onTap))
    .accessibilityElement(children: .ignore)
    .accessibilityLabel(forecast.label)
    .accessibilityValue("\(forecast.temperatureC) degrees")
  }
}

private struct OptionalTap: ViewModifier {
  let action: (() -> Void)?

  @ViewBuilder
  func body(content: Content) -> some View {
    if let action {
      content.onTapGesture(perform: action)
    } else {
      content
    }
  }
}

private let ICON_SIZE: CGFloat = 10
private let DROP_SIZE: CGFloat = 7
private let FONT_SIZE: CGFloat = 11
private let RAIN_FONT_SIZE: CGFloat = 8
